### PR TITLE
Adding UsersResource and resolving issues from ResultsResource

### PR DESCRIPTION
**Pull request by VanDavv**
Here's the Users Resource, tested (more tests soon) and fixes to ResultsResource. I will update this PR, because manual tests are in progress by ArKa, but i think that everything should be fine and ready to deploy

### CODE REVIEW

File: backend/db/src/main/java/com/bls/patronage/db/mappers/UserWithoutPasswordMapper.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.bls.patronage.db.mappers;
```
Comment (mlotysz) [inline backend/db/src/main/java/com/bls/patronage/db/mappers/UserWithoutPasswordMapper.java]: Change name to singular.

File: backend/app/src/test/java/com/bls/patronage/resources/UsersResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        userURI = UriBuilder.fromResource(UsersResource.class).build().toString();
+    }
+
+    @After
+    public void tearDown() {
+        reset(dao);
+    }
+
+    @Test
+    public void createFlashcard() {
```
Comment (mlotysz) [inline backend/app/src/test/java/com/bls/patronage/resources/UsersResourceTest.java]: Name?

File: backend/app/src/test/java/com/bls/patronage/resources/UsersResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        assertThat(response.getStatusInfo()).isEqualTo(Response.Status.OK);
+        verify(dao).createUser(userCaptor.capture());
+        assertThat(userCaptor.getValue().getId()).isNotNull();
+        assertThat(userCaptor.getValue().getEmail()).isEqualTo(userRepresentation.getEmail());
+        assertThat(userCaptor.getValue().getName()).isEqualTo(userRepresentation.getName());
+        assertThat(userCaptor.getValue().getPassword()).isEqualTo(userRepresentation.getPassword());
+    }
+
+    @Test
+    public void createFlashcardWithoutQuestionAndAnswer() {
```
Comment (mlotysz) [inline backend/app/src/test/java/com/bls/patronage/resources/UsersResourceTest.java]: And name?

File: backend/app/src/test/java/com/bls/patronage/resources/UserResourceTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                .request(MediaType.APPLICATION_JSON_TYPE)
+                .delete();
+
+        assertThat(response.getStatusInfo()).isEqualTo(Response.Status.NO_CONTENT);
+    }
+
+    @Test
+    public void getUser() {
+        when(dao.getUserById(user.getId())).thenReturn(user);
+        final User recievedFlashcard = resources.client().target(userURI)
```
Comment (mlotysz) [inline backend/app/src/test/java/com/bls/patronage/resources/UserResourceTest.java]: recievedFlashcard?

File: backend/app/src/main/java/com/bls/patronage/resources/UsersResource.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    @POST
+    public User createUser(@Valid UserRepresentation user) {
+        User createdUser = new User(UUID.randomUUID(), user.getEmail(), user.getName(), user.getPassword());
+        userDAO.createUser(createdUser);
+        return createdUser;
+    }
+
+    @GET
+    public List<UserWithoutPassword> listUsers() {
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/UsersResource.java]: Remove this method. Nobody should get all our users that easy! :)

File: backend/app/src/main/java/com/bls/patronage/resources/UserResource.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @GET
+    public User getUser(
+            @Valid @PathParam("userId") UUIDParam userId) {
+        return userDAO.getUserById(userId.get());
+    }
+
+    @DELETE
+    public void deleteUser(
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/UserResource.java]: Who i gonna to DELETE user? Anybody? For now - delete this method. Check User Story who could delete user from our system. If yes we need to make sure we delete all user realated data too...

File: backend/app/src/main/java/com/bls/patronage/resources/UserResource.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    private final UserDAO userDAO;
+
+    public UserResource(UserDAO userDAO) {
+        this.userDAO = userDAO;
+    }
+
+    @GET
+    public User getUser(
+            @Valid @PathParam("userId") UUIDParam userId) {
+        return userDAO.getUserById(userId.get());
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/UserResource.java]: Never send out internal password representation. This should return UserWithoutPassword too.

File: backend/app/src/main/java/com/bls/patronage/resources/UserResource.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    public void deleteUser(
+            @Valid @PathParam("userId") UUIDParam userId) {
+        userDAO.deleteUser(userDAO.getUserById(userId.get()).getId());
+    }
+
+    @PUT
+    public void updateUser(
+            @Valid @PathParam("userId") UUID userId,
+            @Valid UserRepresentation user) {
+        User updatedUser = new User(userId, user.getEmail(), user.getName(), user.getPassword());
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/UserResource.java]: And never ever store user password in clear text. Read and implement: https://codahale.com/how-to-safely-store-a-password/ and _then_ http://www.mindrot.org/projects/jBCrypt/. And in previous Patronage I implemented this already... :)

**Comment by mlotysz** [comment]
Few of big bugs are here.
Couple of small copy and paste bugs - if you see you are copying all the time its not good job. There is a better way to implement tests! Find out another way.

File: backend/app/src/main/java/com/bls/patronage/resources/UserResource.java (hunk 8 above)
Comment (VanDavv) [inline backend/app/src/main/java/com/bls/patronage/resources/UserResource.java]: I wanted to do this, but I fought, that we could operate only on hashed passwords. This means, that it's up to our clients to encrypt the password, and, even if someone could catch the response, he wont be able to decode it. But probably it's wrong thinking

Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/UserResource.java]: And that's why we need to have HTTPS in place. (We configure nginx for that and only when we'll got good certificates, we'll see.)

File: backend/app/src/test/java/com/bls/patronage/resources/UsersResourceTest.java (hunk 3 above)
Comment (VanDavv) [inline backend/app/src/test/java/com/bls/patronage/resources/UsersResourceTest.java]: Oh my... I completly forgot to change it, I'm sorry!

File: backend/app/src/main/java/com/bls/patronage/resources/UserResource.java (hunk 6 above)
Comment (VanDavv) [inline backend/app/src/main/java/com/bls/patronage/resources/UserResource.java]: There's nothing about it in the User Story, so i'm deleting this method

File: backend/app/src/main/java/com/bls/patronage/resources/UserResource.java (hunk 7 above)
Comment (VanDavv) [inline backend/app/src/main/java/com/bls/patronage/resources/UserResource.java]: But how will FE know if the user typed the correct password? Or is it our job and they are just getting status codes as an answer?

File: backend/app/src/main/java/com/bls/patronage/resources/UsersResource.java (hunk 5 above)
Comment (VanDavv) [inline backend/app/src/main/java/com/bls/patronage/resources/UsersResource.java]: Removed :)

File: backend/app/src/main/java/com/bls/patronage/resources/UserResource.java (hunk 7 above)
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/UserResource.java]: Yes. HTTP 4xx code are for that. You'll add @Auth and then DW will take care of it. But not now. And even then we won't send out password hash. It's our internal secret...

**Comment by mlotysz** [comment]
Rebase onto new master, please. You don't need to use `@Auth` yet. Just make it compile. If it will be too much mess, close this PR and create new one. It's up to you.

**Comment by VanDavv** [comment]
The problem was that I didn't understand how it should work. Now everything is clear. Fixes are going to be committed today :)

**Comment by VanDavv** [comment]
Ok, I'm going do this one more time, from scratch

**Pull request closed without merging**